### PR TITLE
[java] Sparse tensor support

### PR DESCRIPTION
**Pull request by Craigacp**
**Description**:

Adds support for creating and receiving sparse tensors in the ORT Java API.

CSRC and COO tensors as inputs are tested, but there is no op which accepts a block sparse tensor to test. COO tensors are tested as outputs, but there is no op which emits a CSRC or block sparse tensor to test.

**Motivation and Context**
- Why is this change required? What problem does it solve? Request to expose ORT sparse tensor support in Java.

cc @yuslepukhin

### CODE REVIEW

File: java/src/main/java/ai/onnxruntime/OnnxSparseTensor.java
Comment (yuslepukhin) [inline java/src/main/java/ai/onnxruntime/OnnxSparseTensor.java:26]: Just FYI. Both data and the indices are kept in one single buffer. Values first and then aligned indices right after.

File: java/src/main/java/ai/onnxruntime/OnnxSparseTensor.java
Comment (yuslepukhin) [inline java/src/main/java/ai/onnxruntime/OnnxSparseTensor.java:20]: >OnnxSparseTensor [](http://example.com/codeflow?start=19&length=16)

Just a remark. Onnx is a standard. Onnxruntime is an implemenation. Technically, we deal with a Onnxruntime implementation of tensors.

File: java/src/main/java/ai/onnxruntime/OnnxSparseTensor.java
Comment (yuslepukhin) [inline java/src/main/java/ai/onnxruntime/OnnxSparseTensor.java:108]: >getData [](http://example.com/codeflow?start=67&length=7)

In sparse terminology this is either Values (vs Indices) or NNZ.

File: java/src/main/java/ai/onnxruntime/OnnxSparseTensor.java
Comment (yuslepukhin) [inline java/src/main/java/ai/onnxruntime/OnnxSparseTensor.java:763]: Can common things be in a base class? Such as getValues(), getValuesShape(), getIndicesType(), getSparsityType()?

File: java/src/main/java/ai/onnxruntime/OnnxSparseTensor.java
Comment (yuslepukhin) [inline java/src/main/java/ai/onnxruntime/OnnxSparseTensor.java:627]: >getShape [](http://example.com/codeflow?start=18&length=8)

getDenseShape()?

File: java/src/main/java/ai/onnxruntime/OnnxSparseTensor.java
Comment (yuslepukhin) [inline java/src/main/java/ai/onnxruntime/OnnxSparseTensor.java:599]: >BaseSparseTensor [](http://example.com/codeflow?start=32&length=16)

Just curious, why do we need both base class (which is not used fully IMHO) and the interface? Can base class serve as an interface?

File: java/src/main/java/ai/onnxruntime/OnnxSparseTensor.java
Comment (Craigacp) [reply java/src/main/java/ai/onnxruntime/OnnxSparseTensor.java:20]: The `Onnx...` classes map onto concepts in ONNX, and the `Ort...` classes map onto ONNX Runtime specific concepts. I admit the line is a bit fuzzy here because ONNX doesn't define sparse tensor formats, but it does define generic sparse tensors as a thing.

File: java/src/main/java/ai/onnxruntime/OnnxSparseTensor.java
Comment (Craigacp) [reply java/src/main/java/ai/onnxruntime/OnnxSparseTensor.java:108]: I can rename data to values everywhere.

File: java/src/main/java/ai/onnxruntime/OnnxSparseTensor.java
Comment (Craigacp) [reply java/src/main/java/ai/onnxruntime/OnnxSparseTensor.java:763]: Yes, they could be build on construction and passed into the super constructor for `BaseSparseTensor`. I'd need private static methods to compute the various shapes, which would be about as much code as the current way, but it would be slightly more memory efficient. One issue however is that arrays are mutable, so strictly speaking users can modify the array elements which would confuse that sparse tensor. Usually we'd defensively copy them on the accessor which removes the memory efficiency.

File: java/src/main/java/ai/onnxruntime/OnnxSparseTensor.java
Comment (Craigacp) [reply java/src/main/java/ai/onnxruntime/OnnxSparseTensor.java:627]: Sure.

File: java/src/main/java/ai/onnxruntime/OnnxSparseTensor.java
Comment (Craigacp) [reply java/src/main/java/ai/onnxruntime/OnnxSparseTensor.java:599]: The interface + abstract base class is a fairly standard idiom in Java, though I guess it doesn't offer much flexibility here given I intend to seal the interface so it only has the provided implementations. I can refactor it into a public abstract class.

File: java/src/main/native/ai_onnxruntime_OnnxSparseTensor.c
Comment (yuslepukhin) [inline java/src/main/native/ai_onnxruntime_OnnxSparseTensor.c:40]: >checkOrtStatus [](http://example.com/codeflow?start=4&length=14)

Please, remind how checkOrtStatus works. It does record Java exception, but otherwise is happily continuing calling native API. Should this some how stop on error since it may be feeding incorrect values?

File: java/src/main/java/ai/onnxruntime/OnnxSparseTensor.java
Comment (Craigacp) [reply java/src/main/java/ai/onnxruntime/OnnxSparseTensor.java:599]: `SparseTensor` and `OnnxSparseTensor` represent different objects with different uses. The `SparseTensor` is purely Java side, it holds no references to ORT objects or data and is used as the way to prepare data in Java for supplying into ORT, or for working on the output if a sparse tensor is produced from a run call. `SparseTensor` and its implementations could easily live outside the `OnnxSparseTensor` class, and if Java had a similar package to scipy which provided sparse tensors it wouldn't exist at all we'd just depend on the external definitions. `OnnxSparseTensor` holds the references to the ORT objects and memory, but the data only be accessed from Java by making a `SparseTensor` and copying the data out because there aren't any direct accessors into the ORT sparse tensor object. The `OnnxSparseTensor` has a reference to the buffers that came from the `SparseTensor` to make sure they live longer than the `OnnxSparseTensor`, though I'm not actually sure now if it's copying the memory into the `OrtValue` or referencing it.

`OnnxTensorLike` is a made up class which exists because both `OnnxTensor` and `OnnxSparseTensor` can be inputs to a run call, but `OnnxSequence` and `OnnxMap` cannot. So I need something in the type hierarchy that lets me distinguish them statically.

Comment (Craigacp) [reply java/src/main/java/ai/onnxruntime/OnnxSparseTensor.java:599]: Ooops, no wait, you're right, I'm missing the copy in `OnnxSparseTensor.getDataBuffer()`, `OnnxSparseTensor.getIndicesBuffer()` and `OnnxSparseTensor.getInnerIndicesBuffer()`. I'll fix that.

File: java/src/main/native/ai_onnxruntime_OnnxSparseTensor.c
Comment (Craigacp) [reply java/src/main/native/ai_onnxruntime_OnnxSparseTensor.c:40]: All the JNI code needs revising as the exception isn't necessarily propagated until the next call back into the JVM. So `checkOrtStatus` needs to be modified to return something that can be checked and all the places where it is called need to be modified with extra control flow to allow an early return, potentially several layers deep. I've not done that in this PR as it's going to be a large and confusing change which I'd like to keep separate from this one.

File: java/src/main/java/ai/onnxruntime/OnnxSparseTensor.java
Comment (yuslepukhin) [reply java/src/main/java/ai/onnxruntime/OnnxSparseTensor.java:108]: I think it would a good thing.

File: java/src/main/java/ai/onnxruntime/OnnxSparseTensor.java
Comment (yuslepukhin) [reply java/src/main/java/ai/onnxruntime/OnnxSparseTensor.java:763]: I see. I will leave it up to you, but your comment to that effect would be great.

File: java/src/main/native/ai_onnxruntime_OnnxSparseTensor.c
Comment (yuslepukhin) [reply java/src/main/native/ai_onnxruntime_OnnxSparseTensor.c:40]: I think we are going to SEGSEV if OrtTensorTypeAndShapeInfo is some stray pointer that was not properly populated because something errored out. The only thing we can do is to check for nullptr which we do not do since the pre-req is a valid instance. And if it is not nullptr then we can only assume it is valid.

Comment (Craigacp) [reply java/src/main/native/ai_onnxruntime_OnnxSparseTensor.c:40]: Ok, I'll work on a separate PR for the native fixes.

File: java/src/main/java/ai/onnxruntime/OnnxSparseTensor.java
Comment (Craigacp) [reply java/src/main/java/ai/onnxruntime/OnnxSparseTensor.java]: I've fixed the get methods so they copy the buffer. Once this has merged in I plan to add methods to both `OnnxSparseTensor` and `OnnxTensor` which let users copy the buffer into one they supply, and hopefully that will let me refactor and reduce the get logic a little (by pushing it into `OnnxTensorLike`).

File: java/src/main/java/ai/onnxruntime/OnnxSparseTensor.java
Comment (Craigacp) [reply java/src/main/java/ai/onnxruntime/OnnxSparseTensor.java]: I've moved most of those methods onto the base class. `getSparsityType` is still on the subclasses, I find it a little easier to read there as it's a constant for those subtypes. Makes it harder to get confused.

File: java/src/main/java/ai/onnxruntime/OnnxSparseTensor.java
Comment (Craigacp) [reply java/src/main/java/ai/onnxruntime/OnnxSparseTensor.java]: Done.

**Comment by yuslepukhin** [comment]
One question. Does this code still copy memory for input/output? We should not copy at least for input and if so I believe there is a way to prevent that on output as well.

**Comment by Craigacp** [comment]
It does not copy on input if the buffers are direct. If they aren't direct then they need to be copied into direct ones, which are passed in to the OrtValue. Then the `OnnxSparseTensor` holds a reference to the buffers so they don't get garbage collected.

For outputs we could wrap the bare pointers in direct byte buffers, but if those buffers are exposed out of the `OnnxSparseTensor` then they need to not live past the lifetime of the sparse tensor, and that's tricky to enforce. Currently they copy the direct buffer into another buffer before wrapping the copies in the `COOTensor/CSRCTensor/BlockSparseTensor`.

**Comment by yuslepukhin** [comment]
Well, we are doing just find in C#, holding to the native pointers received from ORT and destroy then when IDisposable is invoked, similar to AutoClosable.
If that does not work, here is idea. If output shapes are known, pre-allocate output buffers, create OrtValues and pass them as outputs with corresponding name mappings.

---
In reply to: [1054582742](https://github.com/microsoft/onnxruntime/pull/10653#issuecomment-1054582742) [](http://example.com/codeflow?ancestors=1054582742,1054582742)

**Comment by Craigacp** [comment]
The direct byte buffer abstraction in Java is different from the native pointer abstraction in C#, and I think that there is not a good solution here for outputs that has a zero copy output path and that also exposes the buffers to users. We could alternatively expose direct access into the OrtValue by implementing chunks of the buffer API on top of `OnnxSparseTensor` (or in private lifetime limited objects inside `OnnxSparseTensor`), but that's a substantial development effort. If you want to use a sparse tensor output from one ORT session and pass it into another ORT session then that's zero copy, it's only if you want to get the values back out into Java that it triggers a copy. When the [Java foreign memory API](http://openjdk.java.net/jeps/419) is completed (and ORT is updated to use that minimum Java version, which might involve forking the Java API from the Android one) then there will be better options here for managing the lifetimes of these chunks of memory.

Pre-allocation could work, but that would require a rewrite of all the output processing in both the Java and the native code. I'd missed the update to `Run` which switched the output pointer over so it was `_Inout_`, that seems to have changed after the Java API was initially integrated. Presumably it throws an exception if one of the supplied `OrtValue`s is the wrong size or shape?

**Comment by Craigacp** [comment]
I fixed the build error. Clang must do different exhaustiveness checking on enums than gcc or msvc as the latter two complained I wasn't assigning a value in a switch which covered all the enum cases. I added a default branch which lines up with `ORT_SPARSE_UNDEFINED` which should make it happy. Plus I rebased after the merge of #10670.

**Comment by yuslepukhin** [comment]
> Pre-allocation could work, but that would require a rewrite of all the output processing in both the Java and the native code. I'd missed the update to `Run` which switched the output pointer over so it was `_Inout_`, that seems to have changed after the Java API was initially integrated. Presumably it throws an exception if one of the supplied `OrtValue`s is the wrong size or shape?

Yes, it throws and it does it soon. The API docs and the parameter annotations still need work. This feature was available for a long time. Let's fix error reporting first and then we can work on copy elimination. Both are very important.

**Review by yuslepukhin — APPROVED** [review]

**Comment by Craigacp** [comment]
Can we get this integrated now the 1.11 release has happened? I can rebase it on master if necessary, then it'll be easier to work on the native binding improvements that we discussed.

**Comment by yuslepukhin** [comment]
> Can we get this integrated now the 1.11 release has happened? I can rebase it on master if necessary, then it'll be easier to work on the native binding improvements that we discussed.

I do not mind we rebase it and merge it.

What is your plan with regards to error reporting work?

**Comment by Craigacp** [comment]
> > Can we get this integrated now the 1.11 release has happened? I can rebase it on master if necessary, then it'll be easier to work on the native binding improvements that we discussed.
> 
> I do not mind we rebase it and merge it.
> 
> What is your plan with regards to error reporting work?

Once this has been merged in I'll start making smaller PRs for the different JNI files so it's not as unpleasant to review. The initial one will be bigger because it'll have the error handling changes in `OrtJniUtil.c` and things called from the session run method. All the rest should be smaller and simpler.

**Comment by Craigacp** [comment]
Is it possible to get this branch merged in? I've fixed the merge conflict, and the jni rewrite I've been doing started from this branch.

**Comment by Craigacp** [comment]
The linter seems to be treating my C code as C++ code, as it's complaining I'm using C style casts. Aside from that issue, is there a code formatter configuration I can get to run over the JNI code? I can use IDE plugins for VS Code or CLion, but I don't have access to Visual Studio.

**Comment by Craigacp** [comment]
I've started work on updating this branch with the new error handling code, but it's not ready for review yet.

**Comment by Craigacp** [comment]
This is ready for review now, I've finished updating it with the new JNI error handling.

File: java/src/main/native/ai_onnxruntime_OnnxSparseTensor.c
Comment (yuslepukhin) [inline java/src/main/native/ai_onnxruntime_OnnxSparseTensor.c:66]: >(OrtValue* [](http://example.com/codeflow?start=65&length=10)

I believe the api takes in const OrtValue*. Same is above for non-mutating interfaces

File: java/src/main/native/ai_onnxruntime_OnnxSparseTensor.c
Comment (yuslepukhin) [inline java/src/main/native/ai_onnxruntime_OnnxSparseTensor.c:232]: dimensions leak

File: java/src/main/native/ai_onnxruntime_OnnxSparseTensor.c
Comment (yuslepukhin) [inline java/src/main/native/ai_onnxruntime_OnnxSparseTensor.c:240]: >dimensions = NULL; [](http://example.com/codeflow?start=2&length=18)

redundant

File: java/src/main/native/ai_onnxruntime_OnnxSparseTensor.c
Comment (yuslepukhin) [inline java/src/main/native/ai_onnxruntime_OnnxSparseTensor.c:305]: dimensions leak

File: java/src/main/native/ai_onnxruntime_OnnxSparseTensor.c
Comment (yuslepukhin) [inline java/src/main/native/ai_onnxruntime_OnnxSparseTensor.c:352]: dimensions leak

**Review by yuslepukhin — CHANGES_REQUESTED** [review]
:clock1:

**Comment by Craigacp** [comment]
This should be ready for review again.

**Comment by Craigacp** [comment]
@yuslepukhin would you mind reviewing this again?

**Comment by yuslepukhin** [comment]
Copying or not, we want to make sure that the output tensors are deallocated when no longer needed.

---
In reply to: [1054694679](https://github.com/microsoft/onnxruntime/pull/10653#issuecomment-1054694679) [](http://example.com/codeflow?ancestors=1054694679,1054582742)

**Comment by yuslepukhin** [comment]
>   public static final int MAX_DIMENSIONS = 8;

We did some profiling, the max dim we ever hit was 5.

---
Refers to: java/src/main/java/ai/onnxruntime/TensorInfo.java:15 in 339f78f. [](commit_id = 339f78f59b38d0055cb07b1cd8e005fe19898019, deletion_comment = False)

File: java/src/main/java/ai/onnxruntime/OnnxSparseTensor.java
Comment (yuslepukhin) [inline java/src/main/java/ai/onnxruntime/OnnxSparseTensor.java:53]: >Construct a COO or block sparse tensor. [](http://example.com/codeflow?start=5&length=39)

I notice that the same code is used with COO and blocksparse tensors. However, it is not clear how they can be compatible.
COO may have either 1-D indices or 2-D. And blocksparse is always 2-D.

File: java/src/main/java/ai/onnxruntime/OnnxSparseTensor.java
Comment (yuslepukhin) [inline java/src/main/java/ai/onnxruntime/OnnxSparseTensor.java:244]: >p>These are the outer indices if it's a CSRC sparse tensor. [](http://example.com/codeflow?start=6&length=59)

I would expect to get all the indices to be able to do anything meaningful with them. Why inner indices are not retrieved? I think this interface is not well thought out.

File: java/src/main/java/ai/onnxruntime/OnnxSparseTensor.java
Comment (yuslepukhin) [inline java/src/main/java/ai/onnxruntime/OnnxSparseTensor.java:388]: > Gets the shape of the (outer) indices. [](http://example.com/codeflow?start=4&length=39)

The name of the method does not match its description

File: java/src/main/java/ai/onnxruntime/OnnxSparseTensor.java
Comment (yuslepukhin) [inline java/src/main/java/ai/onnxruntime/OnnxSparseTensor.java:610]: > indicesShape; [](http://example.com/codeflow?start=24&length=14)

CSRC has two indices, each with its own shape.

Comment (yuslepukhin) [inline java/src/main/java/ai/onnxruntime/OnnxSparseTensor.java:610]: >] indicesShape; [](http://example.com/codeflow?start=23&length=15)

Would it not be better to have an array of shapes?

File: java/src/main/java/ai/onnxruntime/OnnxTensor.java
Comment (yuslepukhin) [inline java/src/main/java/ai/onnxruntime/OnnxTensor.java:90]: >@Override [](http://example.com/codeflow?start=2&length=9)

Is this a public interface?

File: java/src/test/java/ai/onnxruntime/SparseTensorTest.java
Comment (yuslepukhin) [inline java/src/test/java/ai/onnxruntime/SparseTensorTest.java:28]: >"/generic_sparse_to_dense_matmul.onnx [](http://example.com/codeflow?start=39&length=37)

There is an existing model that is used in the unit tests for the same purpose. It would be good to verify Java interface works the same way. Also we usually check in the script that generates the model.

**Comment by Craigacp** [comment]
> Copying or not, we want to make sure that the output tensors are deallocated when no longer needed.
> 
> In reply to: [1054694679](https://github.com/microsoft/onnxruntime/pull/10653#issuecomment-1054694679)

`OnnxSparseTensor` is closeable, and deallocates the native memory on close. The `SparseTensor` class is all Java side and the `ByteBuffer`s it contains are managed by the garbage collector.

File: java/src/main/java/ai/onnxruntime/OnnxTensor.java
Comment (Craigacp) [reply java/src/main/java/ai/onnxruntime/OnnxTensor.java:90]: I'm unsure what you mean. This comment is attached to the override annotation on the `toString` method, which `OnnxTensor` inherits from `Object` like all Java objects.

File: java/src/main/java/ai/onnxruntime/OnnxSparseTensor.java
Comment (Craigacp) [reply java/src/main/java/ai/onnxruntime/OnnxSparseTensor.java:610]: We don't have a Java side shape class, and all the relevant methods expect a `long[]` as the shape.

File: java/src/main/java/ai/onnxruntime/OnnxSparseTensor.java
Comment (Craigacp) [reply java/src/main/java/ai/onnxruntime/OnnxSparseTensor.java:53]: The factory methods handle that, this is just to get the pointers through into the right place. In both the COO and block sparse cases there is only a single pointer to the indices, which is what this cares about.

File: java/src/main/java/ai/onnxruntime/OnnxSparseTensor.java
Comment (Craigacp) [reply java/src/main/java/ai/onnxruntime/OnnxSparseTensor.java:244]: They are accessed in two different ways for the CSRC tensor, as there are two methods that return them in the C API. This mirrors that into Java. We can't easily have the Java code return different numbers of things dependent on the native state, so in the CSRC case users need to do two calls to get the two buffers.

File: java/src/main/java/ai/onnxruntime/OnnxSparseTensor.java
Comment (Craigacp) [reply java/src/main/java/ai/onnxruntime/OnnxSparseTensor.java:388]: In the case of block sparse then this is `[num_blocks, block_shape1, block_shape2,...]`, in the case of coo this is `[numNonZero, dimension]` or `[numNonZero]`, and in the case of CSRC it's `[numRows+1]`.  I'm not sure why this requires returning an array of arrays. In the case of the block sparse it is packed, but that's how it's stored in the native code.

I agree if there are more sparse formats then it will need some more work, but as I don't know what the other possible sparse formats are then I'm not sure how to make it more flexible.

File: java/src/test/java/ai/onnxruntime/SparseTensorTest.java
Comment (Craigacp) [reply java/src/test/java/ai/onnxruntime/SparseTensorTest.java:28]: I added `generic_sparse_to_dense_matmul.onnx` to remove the size restriction on the inputs and outputs in the existing `sparse_to_dense_matmul.onnx` which allowed me to test multiple different indices structures without making more models. I honestly don't remember how I created this model, it might have been at the jshell prompt writing the protobuf directly. I can make a static method on `SparseTensorTest` that creates it if you want.

The one that produces a sparse tensor as an output is `sparse_initializer_as_output.onnx` and I use that later in this test file. If there are more models which I should be using to test this code let me know, I couldn't spot any others.

File: java/src/main/java/ai/onnxruntime/OnnxTensor.java
Comment (yuslepukhin) [reply java/src/main/java/ai/onnxruntime/OnnxTensor.java:90]: What I meant is getInfo() used to be a public interface. Is it Ok to remove()

File: java/src/main/java/ai/onnxruntime/OnnxSparseTensor.java
Comment (yuslepukhin) [reply java/src/main/java/ai/onnxruntime/OnnxSparseTensor.java:610]: I would make a Index class with shape inside and all other relevant information. Then we would have no semantic overloading. SparseTensor would always contain what is needed. Alternatively, we could store Tensor classes in the indices array. Each tensor can have a shape and indices data. In C++ we simply project it to the buffer, here we can hold it.

File: java/src/main/java/ai/onnxruntime/OnnxTensor.java
Comment (Craigacp) [reply java/src/main/java/ai/onnxruntime/OnnxTensor.java:90]: Ah, I moved `getInfo` up to `OnnxTensorLike` as it's shared between the dense and sparse tensors.

File: java/src/main/java/ai/onnxruntime/OnnxSparseTensor.java
Comment (Craigacp) [reply java/src/main/java/ai/onnxruntime/OnnxSparseTensor.java:610]: It's not accessible like that through the C API though, and that's what's represented here. My goal was to make something that was as transparent as possible, because there isn't a well accepted sparse tensor library in Java that we can bind to, so I figure people who want sparse behaviour will have created their own and be frustrated by any abstraction I try to layer in there.

**Review by yuslepukhin — APPROVED** [review]

**Comment by yuslepukhin** [comment]
Please, rebase to the latest to make sure we have all the pipeline fixes.

**Comment by Craigacp** [comment]
Done.

**Comment by Craigacp** [comment]
I wish `clang` had the same processing code for deciding definite assignment through switch as `gcc` and `msvc`. It's definitely right, in that the variable was always assigned, but someone needs to explain that to `gcc`. I've fixed the things `gcc` complained about, so you'll need to kick off the pipelines again.

**Review by yuslepukhin — APPROVED** [review]

**Pull request merged**
